src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java

```diff
@@ -0,0 +1,392 @@
+package org.broadinstitute.hellbender.utils.pileup;
+
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMReadGroupRecord;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.BaseUtils;
+import org.broadinstitute.hellbender.utils.fragments.FragmentCollection;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a pileup of reads at a given position.
+ */
+public final class ReadPileup implements Iterable<PileupElement>{
+    private final Locatable loc;
+    private final List<PileupElement> pileupElements;
+
+    /**
+     * Create a new version of a read backed pileup at loc, using the reads and their corresponding
+     * offsets.
+     * Note: This internal-use constructor keeps an alias pointer to the given list.
+     */
+    @VisibleForTesting
+    ReadPileup(final Locatable loc, final List<PileupElement> pileup) {
+        this.loc = loc;
+        this.pileupElements = pileup;
+    }
+
+    /**
+     * Create a new pileup without any aligned reads
+     */
+    public ReadPileup(final Locatable loc) {
+        this(loc, Collections.emptyList());
+    }
+
+    /**
+     * Create a new pileup with the given reads.
+     */
+    public ReadPileup(final Locatable loc, final List<GATKRead> reads, final int offset) {
+        this(loc, readsOffsets2Pileup(reads, offset));
+    }
+
+
+    /**
+     * Create a new pileup with the given reads.
+     */
+    public ReadPileup(final Locatable loc, final List<GATKRead> reads, final List<Integer> offsets) {
+        this(loc, readsOffsets2Pileup(reads, offsets));
+    }
+
+    /**
+     * Helper routine for converting reads and offset lists to a PileupElement list.
+     */
+    private static List<PileupElement> readsOffsets2Pileup(final List<GATKRead> reads, final List<Integer> offsets) {
+        if (reads == null) {
+            throw new GATKException("Illegal null read list in UnifiedReadBackedPileup");
+        }
+        if (offsets == null) {
+            throw new GATKException("Illegal null offsets list in UnifiedReadBackedPileup");
+        }
+        if (reads.size() != offsets.size()) {
+            throw new GATKException("Reads and offset lists have different sizes!");
+        }
+
+        final List<PileupElement> pileup = new LinkedList<>();
```

Comment:
Prefer `ArrayList` to `LinkedList` (better data locality). Specify an initial size equal to `reads.size()`